sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/DoubleBase2ExponentialHistogramAggregator.java

```diff
@@ -131,15 +143,18 @@ protected synchronized ExponentialHistogramPointData doAggregateThenMaybeReset(
     }
 
     private static ExponentialHistogramBuckets resolveBuckets(
-        @Nullable DoubleBase2ExponentialHistogramBuckets buckets, int scale, boolean reset) {
+        @Nullable DoubleBase2ExponentialHistogramBuckets buckets,
+        MutableExponentialHistogramBuckets mutableBuckets,
+        int scale,
+        boolean reset) {
       if (buckets == null) {
         return EmptyExponentialHistogramBuckets.get(scale);
       }
-      ExponentialHistogramBuckets copy = buckets.copy();
+      mutableBuckets.set(buckets);
       if (reset) {
         buckets.clear();
       }
-      return copy;
+      return mutableBuckets;
```

Comment:
This was one idea I entertained for awhile performance gain.  How are you avoiding multi-threads touching this data?

Is it because you're only returning this to ONE metric-reader at a time and the "hot path" of writes is still writing to the underlying data allocated in this handle?

If so, VERY clever. We should document this in the handle class how it works and why it's safe.

Reply:
> Is it because you're only returning this to ONE metric-reader at a time and the "hot path" of writes is still writing to the underlying data allocated in this handle?

Yes exactly. While we support multiple readers, we don't support concurrent reads. As long as readers don't hold on to references to MetricData and try to read after they're done reading, they shouldn't get any weird behavior. Right now this won't work with multiple readers since once PeriodicMetricReader [calls MetricProducer#collectAllMetrics()](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/PeriodicMetricReader.java#L156), another reader will be able to start reading and MetricData will be mutated out from under the PeriodicMetricReader. Ouch. But this is solvable by providing readers a way to communicate to MetricProducer that they're done consuming the data. For example, by adjusting `collectAllMetrics` to accept a `CompletableResultCode` which the reader completes when finished consuming the data, i.e. `MetricProducer#collectAllMetrics(CompleteableResultCode)`. 

As you noticed, this also relies on different objects for writes vs. reads (writes use AggregationHandle, reads use some some mutuable MetricData).

Reply:
Scratch that part about readers needing to communicate when they're finished consuming the data. Each reader has its own copies of metric storage, and the mutable MetricData, so its much simpler: It should be safe as long as a MetricReader doesn't hold on to the MetricData references and try to consume them during a subsequent collect.